### ISSUE TITLE
#738 breaks a handful of our tests

### ISSUE THREAD

**Issue by cpovirk** [eisop/checker-framework#757]
```
  found   : @NonNull Foo<capture#02 extends @NonNull Foo<capture#02>>
  required: @Nullable Foo<? extends @NonNull Object>
```

It looks to me like the `found` type meets the requirements of the `required` type, so I'm not sure exactly where the problem lies.

I'll try to minimize that in the next couple days so that I can send you an actual repro.

**Comment by cpovirk** [comment]
Interestingly, that error arises only(?) in locations that previously had _this_ error, which we had suppressed:

`error: [type.invalid.conflicting.annos] invalid type: conflicting annotations [@NonNull, @Nullable] in type "@NonNull @Nullable Foo<capture#02 extends @NonNull @Nullable Foo<capture#02>>"`

I think I'm close to a self-contained repro, but I'm posting that little bit for now so that I can close a couple windows I have open about it :)

**Comment by cpovirk** [comment]
> Interestingly, that error arises only(?) in locations that previously had this error, which we had suppressed:

The old error is gone, at least :) So #738 hasn't actually increased the number of errors, just changed the kind of error that we see, undoing our old suppressions.

**Comment by wmdietl** [comment]
@cpovirk Sorry for not commenting on this (since a comment in a document when you filed this).

I cannot reproduce a difference between before #738 and after. I also tried typetools release checker-framework-3.42.0.
I tried your version and the following CF-only version:

````java
import org.checkerframework.checker.nullness.qual.NonNull;
import org.checkerframework.checker.nullness.qual.Nullable;

class Eisop757 {
  @SuppressWarnings("nullness:return.type.incompatible")
  <T extends Object> @NonNull T nullCheck(@Nullable T reference) {
    return reference;
  }

  void foo(Foo<? extends Bar<?>> foo) {
    Bar<?> cell = nullCheck(foo.bar());
  }

  interface Foo<T extends Bar<T>> {
    @Nullable T bar();
  }

  interface Bar<T extends Object> {}
}
````

Are you using the CF Nullness Checker or the JSpecify reference checker?

**Comment by cushon** [comment]
I spent some time trying to minimize the repro, and got something that was similar to https://github.com/eisop/checker-framework/issues/757#issuecomment-2091490663 where the error only showed up on `checker-framework-3.42.0-eisop4`. I was running the CF Nullness Checker with e.g. `./checker-framework-3.42.0-eisop4/checker/bin/javac -processor nullness -cp cp.jar Foo.java`

However when I tried to minimize the repro and deleted seemingly unrelated methods in the same file, the error started showing up for both `checker-framework-3.42.0-eisop3` and `checker-framework-3.42.0-eisop4`.

It seems like something very fragile was happening that effectively suppressed that diagnostic from being emitted, and now it's consistently being reported. I'm not sure it's worth investigating this as a regression.

**Comment by cpovirk** [comment]
Thanks, @cushon!

Over time, my memory of this problem (with the CF Nullness Checker) and the problem I reported with https://github.com/eisop/checker-framework/pull/746 (with the JSpecify reference checker) has gotten fuzzy. Sorry again for not getting back to them earlier, at least to answer the easy question from your May 22 post.

I note that I can reproduce a behavior change between the two versions by passing `-AatfDoNotCache`. (Or at least I was able to just now :))

```
checker-framework-3.42.0-eisop3 $ checker/bin/javac -cp $HOME/.m2/repository/org/jspecify/jspecify/1.0.0/jspecify-1.0.0.jar -processor nullness ~/code/Eisop757.java -AatfDoNotCache
Note: The Checker Framework is tested with JDK 8, 11, 17, and 21. You are using version 22.
/usr/local/google/home/cpovirk/code/Eisop757.java:13: error: [type.invalid.conflicting.annos] invalid type: conflicting annotations [@NonNull, @Nullable] in type "@NonNull @Nullable Bar<capture#01 extends @NonNull @Nullable Bar<capture#01>>"
    Bar<?> cell = nullCheck(foo.bar());
                           ^
1 error
```

```
checker-framework-3.42.0-eisop4 $ checker/bin/javac -cp $HOME/.m2/repository/org/jspecify/jspecify/1.0.0/jspecify-1.0.0.jar -processor nullness ~/code/Eisop757.java -AatfDoNotCache
Note: The Checker Framework is tested with JDK 8, 11, 17, and 21. You are using version 22.
/usr/local/google/home/cpovirk/code/Eisop757.java:13: error: [assignment.type.incompatible] incompatible types in assignment.
    Bar<?> cell = nullCheck(foo.bar());
                           ^
  found   : @NonNull Bar<capture#01 extends @NonNull Bar<capture#01>>
  required: @Nullable Bar<? extends @NonNull Object>
1 error
```

Since this is just a change in the kind of error, we're going to suppress it (as I should have back in May if I'd been thinking about this more). But maybe the `-AatfDoNotCache` information at least gives you a lead on debugging if you want to look into this for your own purposes.

**Comment by wmdietl** [comment]
Thanks for the extra information @cushon and @cpovirk !
The difference when using `-AatfDoNotCache` is indeed interesting.
